connectors/source.py

```diff
@@ -582,6 +586,15 @@ async def get_access_control(self):
         """
         raise NotImplementedError
 
+    @contextmanager
+    def with_error_monitoring(self):
+        try:
+            yield
+            self.error_monitor.track_success()
+        except Exception as ex:
+            self._logger.error(ex)
+            self.error_monitor.track_error(ex)
```

Comment:
Something we had in ent-search was the concept of "fatal exceptions". See: https://github.com/elastic/ent-search/blob/a1ddc884ad1543f61a91da7b9511a01090adcac5/connectors/lib/connectors/content_sources/base/extractor.rb#L132-L139

it might be worth adding that here too, for things like aio CanceledError, OOME, etc, to make sure we fail-fast when necessary.

Reply:
Yeah good call. OOME can't get caught, but stuff like CancelledError + having a custom fatal exception class should help

Reply:
Okay, since we don't have code that uses this I'd address it later once we actually use this code somewhere, for now this doesn't make any difference.

I'd also like to separately add handlers for terminal exceptions - for example our download function and some other do `catch Exception` that can catch and "handle" cancelled errors in the weird way